src/Razor/test/Microsoft.VisualStudio.LanguageServerClient.Razor.Test/DefaultLSPEditorServiceTest.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Utilities;
+using Moq;
+using Xunit;
+using Range = Microsoft.VisualStudio.LanguageServer.Protocol.Range;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    public class DefaultLSPEditorServiceTest
+    {
+        [Fact]
+        public void ExtractsCursorPlaceholder_AppliesEditsCorrectly()
+        {
+            // Arrange
+            var expectedCursorPosition = new Position(2, 10);
+            var snapshot = new TestTextSnapshot($@"
+@{{
+    <text>{LanguageServerConstants.CursorPlaceholderString}</text>
+}}");
+            var edits = new[]
+            {
+                new TextEdit()
+                {
+                    NewText = $"{LanguageServerConstants.CursorPlaceholderString}</text>",
+                    Range = new Range() { Start = expectedCursorPosition, End = expectedCursorPosition },
+                }
+            };
+
+            // Act
+            var cursorPosition = DefaultLSPEditorService.ExtractCursorPlaceholder(snapshot, edits);
+
+            // Assert
+            Assert.Equal(expectedCursorPosition, cursorPosition);
+        }
+
+        [Fact]
+        public void ExtractsCursorPlaceholder_MultipleEdits_AppliesEditsCorrectly()
+        {
+            // Arrange
+            var expectedCursorPosition = new Position(2, 10);
+            var snapshot = new TestTextSnapshot($@"
+@{{
+    <text>{LanguageServerConstants.CursorPlaceholderString}</text>
+}}");
+            var edits = new[]
+            {
+                new TextEdit()
+                {
+                    NewText = $"unrelated Edit",
+                    Range = new Range() { Start = new Position(0, 0), End = new Position(0, 1) },
+                },
+                new TextEdit()
+                {
+                    NewText = $"{LanguageServerConstants.CursorPlaceholderString}</text>",
+                    Range = new Range() { Start = expectedCursorPosition, End = expectedCursorPosition },
+                }
+            };
+
+            // Act
+            var cursorPosition = DefaultLSPEditorService.ExtractCursorPlaceholder(snapshot, edits);
+
+            // Assert
+            Assert.Equal(expectedCursorPosition, cursorPosition);
+        }
+
+        private class TestTextSnapshot : ITextSnapshot
```

Comment:
We have a StringTextSnapshot type in Editor.Razor.Test.Common. We should consider copying that over to our language server test common project